File: BOJ/DFS,BFS/4963.py
# 4963 섬의 개수
# https://www.acmicpc.net/problem/4963

import sys
sys.setrecursionlimit(10000)
def input(): return sys.stdin.readline().rstrip()


def solution(w, h):
    answer = 0
    board = [list(map(int, input().split())) for _ in range(h)]
    visited = [[False] * w for _ in range(h)]
    dir = [[1, 0], [0, 1], [-1, 0], [0, -1],
           [1, -1], [1, 1], [-1, -1], [-1, 1]]

    def DFS(x, y):
        for d in dir:
            nx = x + d[0]
            ny = y + d[1]

            if nx < 0 or ny < 0 or nx >= h or ny >= w:
                continue

            if visited[nx][ny] or board[nx][ny] == 0:
                continue

            visited[nx][ny] = True
            DFS(nx, ny)

    for i in range(h):
        for j in range(w):
            if board[i][j] == 1 and not visited[i][j]:
                answer += 1
                visited[i][j] = True
                DFS(i, j)

    print(answer)
    return


while True:
    w, h = map(int, input().split())

    if w == 0 and h == 0:
        break

    solution(w, h)

File: BOJ/DFS,BFS/4963/4963.cpp
// 4963 섬의 개수
// https://www.acmicpc.net/problem/4963
#include <iostream>
#include <vector>

using namespace std;

int dir[8][2] = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}, 
                {1, 1}, {-1, -1}, {-1, 1}, {1, -1}};

void DFS(vector<vector<int>> board, vector<vector<int>> &visited, int x, int y) {
    visited[x][y] = 1;

    for (auto& d: dir)
    {
        int nx = x + d[0];
        int ny = y + d[1];

        if (nx < 0 || ny < 0 || nx >= board.size() || ny >= board[0].size())
            continue;
        
        if (visited[nx][ny] || !board[nx][ny]) continue;

        DFS(board, visited, nx, ny);
    }
    
    return;
}

int solution(int h, int w, vector<vector<int>> board) {
    int answer = 0;
    vector<vector<int>> visited(h, vector<int> (w, 0));

    for (int i = 0; i < h; ++i)
    {
        for (int j = 0; j < w; ++j)
        {
            if (board[i][j] && !visited[i][j])
            {
                DFS(board, visited, i, j);
                answer++;
            }
        }
    }

    return answer;
}

int main(void) {
    int T, w, h;
    vector<vector<int>> board;

    while (1)
    {
        cin >> w >> h;
        if (w == 0 && h == 0) break;

        board = vector<vector<int>> (h, vector<int>(w, 0));

        for (int i = 0; i < h; ++i)
        {
            for (int j = 0; j < w; ++j)
            {
                cin >> board[i][j];
            }
        }

        cout << solution(h, w, board) << '\n';
    }
    return 0;
}